### PR TITLE
xhb adapter added - use AppNexus test ad unit

### PR DESCRIPTION
**Pull request by daniel-hoffmann**
## Type of change

- [x ] New bidder adapter

- test parameters for validating bids
- same as AppNexus adapter

```
{
  bidder: 'xhb',
  params: {
    // same as AppNexus adapter
  }
}
```
- daniel.hoffmann@xaxis.com
- [x] official adapter submission

### CODE REVIEW

**Comment by mkendall07** [comment]
@daniel-hoffmann 
If this is a copy of `appnexus.js` you should alias the adapter instead. See `adapters.json` for examples.

**Comment by daniel-hoffmann** [comment]
@mkendall07 
It's not a copy and using an alias is not feasible unfortunately, as we made some changes that are required for us to use Prebid. Let me know if you have any questions!

**Comment by mkendall07** [comment]
@daniel-hoffmann 
What are the deltas between AppNexus and this adapter? There is no reason to have so much duplicate code inside an adapter.

Btw, your setting the CPM of all bids to `0.01`.  https://github.com/prebid/Prebid.js/pull/672/files#diff-4d57568d0e9ce5fbaa50e6e0af5b37faR131

**Comment by daniel-hoffmann** [comment]
@mkendall07 
I adjusted the adapter, please take a look and let me know if this is okay.
Is it possible to make sure that it will be included in 0.18.0?
Let me know if you have any questions! :)

File: src/adapters/xhb.js [hunk 1]
```diff
@@ -0,0 +1,187 @@
+import { getBidRequest } from '../utils.js';
+
+var CONSTANTS = require('../constants.json');
```
Comment (mkendall07) [inline src/adapters/xhb.js]: prefer `const` here

File: src/adapters/xhb.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+var utils = require('../utils.js');
+var adloader = require('../adloader.js');
+var bidmanager = require('../bidmanager.js');
+var bidfactory = require('../bidfactory.js');
+//var Adapter = require('./adapter.js');
```
Comment (mkendall07) [inline src/adapters/xhb.js]: can remove

File: src/adapters/xhb.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+var XhbAdapter = function XhbAdapter() {
+
+    function buildJPTCall(bid, callbackId) {
+        //determine tag params
+        var placementId = utils.getBidIdParamater('placementId', bid.params);
```
Comment (mkendall07) [inline src/adapters/xhb.js]: `let` or `const` as appropriate.

File: src/adapters/xhb.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                    sizeQueryString = sizeQueryString.slice(0, sizeQueryString.length - 1);
+                }
+            }
+        }
+
+        if (sizeQueryString) {
+            jptCall += sizeQueryString + '&';
+        }
+
+
```
Comment (mkendall07) [inline src/adapters/xhb.js]: remove extra new lines.

File: src/adapters/xhb.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        //append referrer
+        if (referrer === '') {
+            referrer = utils.getTopWindowUrl();
+        }
+
+        jptCall = utils.tryAppendQueryString(jptCall, 'referrer', referrer);
+        jptCall = utils.tryAppendQueryString(jptCall, 'alt_referrer', altReferrer);
+
+
```
Comment (mkendall07) [inline src/adapters/xhb.js]: remove new lines.

File: src/adapters/xhb.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+
+        //remove the trailing "&"
+        if (jptCall.lastIndexOf('&') === jptCall.length - 1) {
+            jptCall = jptCall.substring(0, jptCall.length - 1);
+        }
+
+        // @if NODE_ENV='debug'
```
Comment (mkendall07) [inline src/adapters/xhb.js]: these code comments don't actually work anymore. You can strip out any blocks with this syntax.

File: src/adapters/xhb.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        utils.logMessage('jpt request built: ' + jptCall);
+        // @endif
+
+        //append a timer here to track latency
+        bid.startTime = new Date().getTime();
```
Comment (mkendall07) [inline src/adapters/xhb.js]: Doesn't appear this is needed anywhere?

File: src/adapters/xhb.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+        return jptCall;
+    }
+
+    //expose the callback to the global object:
+    $$PREBID_GLOBAL$$.handleXhbCB = function (jptResponseObj) {
+
```
Comment (mkendall07) [inline src/adapters/xhb.js]: line break can be removed.

File: src/adapters/xhb.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+            // @if NODE_ENV='debug'
+            utils.logMessage('JSONP callback function called for ad ID: ' + id);
+            // @endif
+
+            var bid = [];
+            if (jptResponseObj.result && jptResponseObj.result.ad && jptResponseObj.result.ad !== '') {
+
+
+                //responseCPM = parseInt(jptResponseObj.result.cpm, 10);
```
Comment (mkendall07) [inline src/adapters/xhb.js]: These commented lines can be removed and all other commented code lines.

File: src/adapters/xhb.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+                //in order to avoid using floats
+                //switch CPM to "dollar/cent"
+                //responseCPM = responseCPM / 10000;
+                responseCPM = 0.00;
+
+                //store bid response
+                //bid status is good (indicating 1)
+
+                var adId = jptResponseObj.result.creative_id;
+                bid = bidfactory.createBid(1);
```
Comment (mkendall07) [inline src/adapters/xhb.js]: Please pass through the original bid request object (`bidObj`) as the 2nd param to `createBid`.
`bid = bidfactory.createBid(CONSTANTS.STATUS.GOOD, bidObj);`

File: src/adapters/xhb.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+                //indicate that there is no bid for this placement
+                bid = bidfactory.createBid(2);
+                bid.bidderCode = bidCode;
+                bidmanager.addBidResponse(placementCode, bid);
+            }
+
+        } else {
+            //no response data
+            // @if NODE_ENV='debug'
+            utils.logMessage('No prebid response for placement %%PLACEMENT%%');
```
Comment (mkendall07) [inline src/adapters/xhb.js]: this else condition can be dropped.

**Comment by mkendall07** [comment]
Yeah I know this was a copy / paste :) 
The copied code was from the original adapter wrote for prebid.js and has a lot of rough edges. Probably just better to clean up this version (the appnexus.js adapter needs to get cleaned up too).

**Comment by daniel-hoffmann** [comment]
Ah, okay. :) Might be possible to just use the ast version for the adapter then - if you think that is the better solution? Otherwise just let's go with this.

**Comment by mkendall07** [comment]
I think better to go with this. We are actively making changes to AST. Thanks

**Comment by daniel-hoffmann** [comment]
Should be fine now - please review again. :)

**Comment by daniel-hoffmann** [comment]
@mkendall07 Sorry, forgot to add the @ - are the changes okay this way? Do you want me to make any additional ones? :)

**Comment by mkendall07** [comment]
@daniel-hoffmann 
I think it's good I just need to test it out quickly. I'll reply in a bit.

**Comment by mkendall07** [comment]
@daniel-hoffmann 
Please add `xhb` into `adapters.json`.

Additionally there is lots of style errors when running `gulp serve` - if you run a `jscs-fixer` type plugin it should automatically fix it. We should be good for merge after that.

**Comment by daniel-hoffmann** [comment]
@mkendall07 Done! :)

**Comment by mkendall07** [comment]
@daniel-hoffmann please see https://github.com/prebid/Prebid.js/pull/938

**Pull request closed without merging**